File: Python_Algorithm/a_star.py
import argparse
import time
import tracemalloc
import heapq
from math import sqrt
from enum import Enum

from read_from_csv import read_from_csv
from reconstruct_path import reconstruct_path
from write_to_json import write_to_json

class HeuristicType(Enum):
    """An enumeration representing different types of heuristics for the A*-algorithm.
    
    Attributes:
        MANHATTAN_DISTANCE: Represents the Manhattan distance.
        AIRPLANE_DISTANCE: Represents the Airplane distance.
    """
    MANHATTAN_DISTANCE = "manhattan"
    AIRPLANE_DISTANCE = "airplane"


def calculate_manhattan_distance(current_position, goal_position):
    """Calculate the Manhattan distance between two tiles.

    Arguments:
        current_position (tuple[int, int]): The current position on the map.
        goal_position (tuple[int, int]): The goal position on the map.

    Returns:
        int: The Manhattan distance between the current position and the goal.
    """
    return abs(current_position[0] - goal_position[0]) + abs(current_position[1] - goal_position[1])


def calculate_airplane_distance(current_position, goal_position):
    """Calculate the Airplane distance between two tiles.

    Arguments:
        current_position (tuple[int, int]): The current position on the map.
        goal_position (tuple[int, int]): The goal position on the map.

    Returns:
        float: The Airplane distance between the current position and the goal.
    """
    return sqrt((current_position[0] - goal_position[0])**2 + (current_position[1] - goal_position[1])**2)

def is_valid_neighbour_tile(row_index, column_index, map):
    """Check if a given tile is part of the map, not an obstacle, and was not visited before.

    Arguments:
        row_index (int): The row index of the current position.
        column_index (int): The columns index of the current position.
        map (list[list[int]]): The map on which the A*-algorithm is applied.

    Returns:
        boolean: Returns whether the tile is a valid neighbour or not.
    """
    return (
        0 <= row_index < len(map)
        and 0 <= column_index < len(map[0])
        and map[row_index][column_index] != 1
    )


def a_star(map, algorithm_map, start_position, goal_position, heuristic_type, status_code):
    """Apply the A* Algorithm.

    Arguments:
        map (list[list[int]]): The map in form of a list of lists of integers.
        algorithm_map (list[list[int]]): The prepared map in form of a list of lists of integers that will be overwritten by the A*-algorithm.
        start_position (tuple[int, int]): The start position on the map for the A*-algorithm.
        goal_position (tuple[int, int]): The goal position on the map for the A*-algorithm.
        heuristic_type (HeuristicType): The mathematical function chosen to calculate the distance from a tile to the goal (Manhattan or Airplane).
        status_code (int): The status code returned when reading the CSV-file.

    Returns:
        list[list[int]]: The map created by the A*-algorithm filled with the determined values.
        int: The status code symbolizing success or a specific error type prior to or while performing the A*-algorithm.
        int: The shortest determined path length in the form of tiles towards the goal.
        list[tuple[int, int]]: The shortest determined path in form of the positions towards the goal.
        float: The time the A*-algorithm needs to finish in seconds.
        float: The memory used during the execution in MB.
    """
    # Checks if an error was encountered prior to the A*-algorithm
    if status_code != 200:
        return algorithm_map, status_code, -1, [], 0.0, 0.0

    # Saves the start_time and the observation of the memory usage is started
    start_time = time.perf_counter()
    tracemalloc.start()

    # Selects the heuristic function based on the heuristic_type argument
    if heuristic_type == HeuristicType.MANHATTAN_DISTANCE.value:
        heuristic = calculate_manhattan_distance
    elif heuristic_type == HeuristicType.AIRPLANE_DISTANCE.value:
        heuristic = calculate_airplane_distance
    else:
        # Returns error code, if wrong heuristic_type argument was given
        return algorithm_map, 401, -1, [], 0.0, 0.0

    # Initializes all the components needed for the A*-algorithm
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    open_priority_queue = []
    previous_positions = {}
    path_length = -1

    # Starts the A*-algorithm at the start position by setting its distance and adding it to the priority queue
    algorithm_map[start_position[0]][start_position[1]] = 0
    heapq.heappush(open_priority_queue, (0, start_position[0], start_position[1], 0))
    f_scores = {start_position: 0}

    # Performs the A*-algorithm as long as there are still tiles in the open_priority_queue with unchecked neighbours
    while open_priority_queue:
        # Reads a position and saved distance to the start position (g_score) from the open_priority_queue based on the lowest estimated distance towards the goal position (f_score)
        f_score, row_index, column_index, g_score = heapq.heappop(open_priority_queue)

        print(f_score, row_index, column_index, g_score)

        # If the read position is the goal position its distance is saved as the path_length and the algorithm stops
        if (row_index, column_index) == goal_position:
            path_length = g_score
            break

        # For each of the four tiles next to the read position the g_score and f_score is saved
        for row_direction, column_direction in directions:
            current_row_index = row_index + row_direction
            current_column_index = column_index + column_direction

            # Prior to saving the g_score and f_score it is checked, if the tile is a valid neighbour
            if is_valid_neighbour_tile(current_row_index, current_column_index, map):
                # The g_score and f_score of the current position is calculated
                current_g_score = g_score + 1
                current_f_score = current_g_score + heuristic((current_row_index, current_column_index), goal_position)

                # The current values only get added to the open_priority_queue if there was no f_score calculated for the position which is lower than the current_f_score
                if current_f_score < f_scores.get((current_row_index, current_column_index), float('inf')):
                    # The determined g_score which represents the distance is saved to the algorithm_map, the f_score gets added to f_scores, and all the values get added to the open_priority_queue
                    algorithm_map[current_row_index][current_column_index] = current_g_score
                    f_scores[current_row_index, current_column_index] = current_f_score
                    heapq.heappush(open_priority_queue, (current_f_score, current_row_index, current_column_index, current_g_score))

                    # Saves the previous position to each current position to determine the path towards the goal at a later step
                    previous_positions[(current_row_index, current_column_index)] = (row_index, column_index)

    # If no path towards the goal could be found, the corresponding status_code gets returned
    if path_length == -1:
        status_code = 404
    else:
        # Reconstruct the A*-algorithms path
        path_positions = reconstruct_path(start_position, goal_position, previous_positions)

    # The memory usage gets saved and the tracking is stopped
    _, peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    # The end_time gets saved and the computing_time in seconds is calculated by subtracting the start_time and rounding the result to two decimal places
    end_time = time.perf_counter()
    computing_time = end_time - start_time

    # The memory usage gets converted to MB
    memory_usage = peak / 10**6

    return algorithm_map, status_code, path_length, path_positions, computing_time, memory_usage


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="A*-Algorithmus auf Basis einer Map in einer CSV-Datei und Ausgabe in einer JSON-Datei")
    parser.add_argument("csv_path", help="Pfad zur Eingabe-CSV-Datei")
    parser.add_argument("json_path", help="Pfad zur Ausgabe-JSON-Datei")
    parser.add_argument(
        "heuristic_type",
        choices=[type.value for type in HeuristicType],
        help="Die heuristische Methode zur Berechnung der Distanz zwischen einzelnen Feldern: 'Manhattan'- oder 'Airplane'-Distanz"
    )
    args = parser.parse_args()

    # Load map and apply A* algorithm
    map, algorithm_map, start_position, goal_position, status_code = read_from_csv(args.csv_path)
    algorithm_map, status_code, path_length, path_positions, computing_time, memory_usage = a_star(map, algorithm_map, start_position, goal_position, args.heuristic_type, status_code)
    write_to_json(args.json_path, algorithm_map, status_code, path_length, path_positions, computing_time, memory_usage)

File: Bahnplanungsalgorithmus/Java_Algorithms/AStar.java
import java.util.Comparator;
import java.util.HashMap;
import java.util.PriorityQueue;
import java.util.function.BiFunction;

/**
 * AStar is the class that applies the A*-algorithm
 */
public class AStar {
    /**
     * The four directions to neighbour tiles
     */
    private static final int[][] DIRECTIONS = {
            {-1, 0}, {1, 0}, {0, -1}, {0, 1}
    };

    /**
     * The map in form of a multidimensional array
     */
    private final int[][] map;

    /**
     * The prepared map in form of a multidimensional array that will be overwritten by the A*-algorithm
     */
    private final int[][] algorithmMap;

    /**
     * The start position row index on the map for the A*-algorithm
     */
    private final int startRowIndex;

    /**
     * The start position column index on the map for the A*-algorithm
     */
    private final int startColumnIndex;

    /**
     * The goal position row index on the map for the A*-algorithm
     */
    private final int goalRowIndex;

    /**
     * The goal position column index on the map for the A*-algorithm
     */
    private final int goalColumnIndex;

    /**
     * The heuristic function type to calculate the distance towards the goal position
     */
    private final String heuristicType;

    /**
     * The status code returned when reading the CSV-file
     */
    private int statusCode;

    /**
     * Constructor of the AStar class
     * @param map The map in form of a multidimensional array
     * @param algorithmMap The prepared map in form of a multidimensional array that will be overwritten by the A*-algorithm
     * @param startRowIndex The start position row index on the map for the A*-algorithm
     * @param startColumnIndex The start position column index on the map for the A*-algorithm
     * @param goalRowIndex The goal position row index on the map for the A*-algorithm
     * @param goalColumnIndex The goal position column index on the map for the A*-algorithm
     * @param heuristicType The heuristic function type to calculate the distance towards the goal position
     * @param statusCode The status code returned when reading the CSV-file
     */
    public AStar (int[][] map, int[][] algorithmMap, int startRowIndex, int startColumnIndex, int goalRowIndex, int goalColumnIndex, String heuristicType, int statusCode) {
        this.map = map;
        this.algorithmMap = algorithmMap;
        this.startRowIndex = startRowIndex;
        this.startColumnIndex = startColumnIndex;
        this.goalRowIndex = goalRowIndex;
        this.goalColumnIndex = goalColumnIndex;
        this.heuristicType = heuristicType;
        this.statusCode = statusCode;
    }

    /**
     * Calculate the Manhattan distance between two tiles
     * @param rowIndex The row index of the current position
     * @param columnIndex The columns index of the current position
     * @return The Manhattan distance between the current position and the goal
     */
    private double calculateManhattanDistance(int rowIndex, int columnIndex) {
        return (Math.abs(rowIndex - goalRowIndex) + Math.abs(columnIndex - goalColumnIndex));
    }

    /**
     * Calculate the Airplane distance between two tiles
     * @param rowIndex The row index of the current position
     * @param columnIndex The columns index of the current position
     * @return The Airplane distance between the current position and the goal
     */
    private double calculateAirplaneDistance(int rowIndex, int columnIndex) {
        return (Math.sqrt(Math.pow((rowIndex - goalRowIndex), 2) + Math.pow((columnIndex - goalColumnIndex), 2)));
    }

    /**
     * Check if a given tile is part of the map, not an obstacle, and was not visited before
     * @param rowIndex The row index of the current position
     * @param columnIndex The columns index of the current position
     * @return Whether the tile is a valid neighbour or not
     */
    private boolean isValidNeighbourTile(int rowIndex, int columnIndex) {
        return (
                (0 <= rowIndex && rowIndex < map.length) &&
                (0 <= columnIndex && columnIndex < map.length) &&
                (map[rowIndex][columnIndex] != 1)
        );
    }

    /**
     * The actual A*-algorithm
     * @return The results of the algorithm in form of an AlgorithmResult
     */
    public AlgorithmResult applyAStar() {
        // Checks if an error was encountered prior to the A*-algorithm
        if (statusCode != 200) {
            return new AlgorithmResult(algorithmMap, statusCode, -1, new int[][]{}, 0, 0);
        }

        // Calls Garbage Collector to clean memory
        System.gc();

        // Saves the startTime and starts the observation of the memory usage
        long startTime = System.nanoTime();
        long memoryBefore = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();

        // Returns error code, if wrong heuristic_type argument was given
        if (!(heuristicType.equalsIgnoreCase(AStarHeuristicType.MANHATTAN_DISTANCE.getValue())|| heuristicType.equalsIgnoreCase(AStarHeuristicType.AIRPLANE_DISTANCE.getValue()))) {
            return new AlgorithmResult(algorithmMap, 401, -1, new int[][]{}, 0, 0);
        } else {
            // Selects the heuristic function based on the heuristicType argument
            BiFunction<Integer, Integer, Double> heuristic;
            if (heuristicType.equalsIgnoreCase(AStarHeuristicType.MANHATTAN_DISTANCE.getValue())) {
                heuristic = this::calculateManhattanDistance;
            } else {
                heuristic = this::calculateAirplaneDistance;
            }

            // Initializes all the components needed for the A*-algorithm
            PriorityQueue<double[]> openPriorityQueue = new PriorityQueue<>(Comparator.comparingDouble((double[] a) -> a[0])
                    .thenComparingDouble(a -> a[1])
                    .thenComparingDouble(a -> a[2])
                    .thenComparingDouble(a -> a[3]));
            HashMap<String, String> previousPositions = new HashMap<>();
            HashMap<String, Integer> gScores = new HashMap<>();
            int pathLength = -1;
            int[][] pathPositions = new int[][]{};

            // Starts the A*-algorithm at the start position by setting its distance and adding it to the priority queue
            algorithmMap[startRowIndex][startColumnIndex] = 0;
            openPriorityQueue.add(new double[]{0, 0, startRowIndex, startColumnIndex});
            gScores.put(startRowIndex + "," + startColumnIndex, 0);

            while (!openPriorityQueue.isEmpty()) {
                double[] position = openPriorityQueue.poll();
                int gScore = (int) position[1];
                int rowIndex = (int) position[2];
                int columnIndex = (int) position[3];

                // If the read position is the goal position its distance is saved as the pathLength and the algorithm stops
                if ((rowIndex == goalRowIndex) && (columnIndex == goalColumnIndex)) {
                    pathLength = gScore;
                    break;
                }

                // For each of the four tiles next to the read position the distance is saved
                for (int[] direction : DIRECTIONS) {
                    int currentRowIndex = rowIndex + direction[0];
                    int currentColumnIndex = columnIndex + direction[1];

                    // Prior to saving the distance it is checked, if the tile is a valid neighbour
                    if (isValidNeighbourTile(currentRowIndex, currentColumnIndex)) {
                        // The gScore of the current position is calculated
                        int currentGScore = gScore + 1;

                        // The current values only get added to the openPriorityQueue if there was no gScore calculated for the position which is lower than the currentGScore
                        if (currentGScore < gScores.getOrDefault(currentRowIndex + "," + currentColumnIndex, Integer.MAX_VALUE)) {
                            // The fScore of the current position is calculated
                            double currentFScore = currentGScore + heuristic.apply(currentRowIndex, currentColumnIndex);

                            // The determined gScore which represents the distance is saved to the algorithmMap, added to gScores, and all the values get added to the openPriorityQueue
                            algorithmMap[currentRowIndex][currentColumnIndex] = currentGScore;
                            gScores.put(currentRowIndex + "," + currentColumnIndex, currentGScore);
                            openPriorityQueue.add(new double[]{currentFScore, currentGScore, currentRowIndex, currentColumnIndex});

                            // Saves the previous position to each current position to determine the path towards the goal at a later step
                            previousPositions.put(currentRowIndex + "," + currentColumnIndex, rowIndex + "," + columnIndex);
                        }
                    }
                }
            }


            // If no path towards the goal could be found, the corresponding status_code gets returned
            if (pathLength == -1){
                statusCode = 404;
            } else {
                // Reconstruct the A*-algorithms path
                PathReconstructor pathReconstructor = new PathReconstructor(startRowIndex, startColumnIndex, goalRowIndex, goalColumnIndex, previousPositions);
                pathPositions = pathReconstructor.reconstructPath();
            }

            // The tracking of the memory usage is stopped
            long memoryAfter = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();

            // The endTime gets saved and the computingTime in seconds is calculated by subtracting the startTime
            long endTime = System.nanoTime();
            double computingTime = (double)(endTime - startTime)/1_000_000_000.0;

            // The memory usage gets converted to MB
            double memoryUsage = (double)(memoryAfter - memoryBefore) / 1_048_576.0;

            return new AlgorithmResult(algorithmMap, statusCode, pathLength, pathPositions, computingTime, memoryUsage);
        }
    }
}